### PR TITLE
document FlowOps extension

### PR DESCRIPTION
**Pull request by rkuhn**

### CODE REVIEW

File: akka-docs/rst/scala/code/docs/stream/TwitterStreamQuickstartDocSpec.scala [hunk 1]
```diff
@@ -32,7 +32,8 @@ object TwitterStreamQuickstartDocSpec {
   //#model
 
   //#tweet-source
-  val tweets: Source[Tweet, NotUsed] //#tweet-source
+  val tweets: Source[Tweet, NotUsed]
+  //#tweet-source
```
Comment (rkuhn) [inline akka-docs/rst/scala/code/docs/stream/TwitterStreamQuickstartDocSpec.scala]: This has been carefully restored, but scalariform will break it with any chance it gets (i.e. during every compilation). We should definitely fix the code formatting automatism …

(in the reformatted form the `//#tweet-source` snippet is not found by sphinx)

Comment (ktoso) [inline akka-docs/rst/scala/code/docs/stream/TwitterStreamQuickstartDocSpec.scala]: I see, could you surround it with formatter:off right away then?

File: akka-docs/rst/scala/code/docs/stream/GraphStageDocSpec.scala [hunk 2]
```diff
@@ -519,4 +519,32 @@ class GraphStageDocSpec extends AkkaSpec {
     sub.cancel()
   }
 
+  "Demonstrate stream extension" when {
+
+    "targeting a Source" in {
+      //#extending-source
+      implicit class SourceDuplicator[Out, Mat](s: Source[Out, Mat]) {
```
Comment (ktoso) [inline akka-docs/rst/scala/code/docs/stream/GraphStageDocSpec.scala]: Should our example extend AnyVal?

File: akka-docs/rst/scala/stream/stream-customize.rst [hunk 3]
```diff
@@ -448,3 +448,37 @@ stage as state of an actor, and the callbacks as the ``receive`` block of the ac
...
+combinator to integrate it into your stream topology. This works with all :class:`FlowOps` sub-types, including the
+ports that you connect with the graph DSL.
+
+Advanced Scala users may wonder whether it is possible to write extension methods that enrich :class:`FlowOps` to
+allow nicer syntax. The short answer is that Scala 2 does not support this in a fully generic fashion, the problem is
+that it is impossible to abstract over the kind of stream that is being extended because :class:`Source`, :class:`Flow`
+and :class:`SubFlow` differ in the number and kind of their type parameters. While it would be possible to write
+an implicit class that enriches them generically, this class would require explicit instantiation with all type
+parameters due to `SI-2712 <https://issues.scala-lang.org/browse/SI-2712>`_. For a partial workaround that unifies
+extensions to :class:`Source` and :class:`Flow` see `this sketch by R. Kuhn <https://gist.github.com/rkuhn/2870fcee4937dda2cad5>`_.
```
Comment (ktoso) [inline akka-docs/rst/scala/stream/stream-customize.rst]: Yeah this is enough of explaining of the crazy way, thanks

File: akka-docs/rst/scala/stream/stream-customize.rst [hunk 4]
```diff
@@ -448,3 +448,37 @@ stage as state of an actor, and the callbacks as the ``receive`` block of the ac
...
   internal state of custom stages because such access can be concurrent with the provided callbacks, leading to undefined
   behavior.
+
+Extending Flow Combinators with Custom Operators
+================================================
+
+The most general way of extending any :class:`Source`, :class:`Flow` or :class:`SubFlow` (e.g. from ``groupBy``) is
+demonstrated above: create a graph of flow-shape like the :class:`Duplicator` example given above and use the ``.via(...)``
+combinator to integrate it into your stream topology. This works with all :class:`FlowOps` sub-types, including the
+ports that you connect with the graph DSL.
```
Comment (ktoso) [inline akka-docs/rst/scala/stream/stream-customize.rst]: Good

**Comment by ktoso** [comment]
LGTM, I'm not sure about explicitly stating the part about stack overflows and giving up

**Comment by drewhk** [comment]
LGTM

File: akka-docs/rst/scala/stream/stream-customize.rst [hunk 5]
```diff
@@ -448,3 +448,37 @@ stage as state of an actor, and the callbacks as the ``receive`` block of the ac
...
+
+Advanced Scala users may wonder whether it is possible to write extension methods that enrich :class:`FlowOps` to
+allow nicer syntax. The short answer is that Scala 2 does not support this in a fully generic fashion, the problem is
```
Comment (analytically) [inline akka-docs/rst/scala/stream/stream-customize.rst]: Scala 2?

Comment (ktoso) [inline akka-docs/rst/scala/stream/stream-customize.rst]: That's what you're currently using ;)
Scala 2.x

**Comment by ktoso** [comment]
Thanks @rkuhn !

**Comment by patriknw** [comment]
LGTM, thanks @rkuhn

**Pull request merged**